pkg/server/datastore/sqlstore/sqlstore_test.go

```diff
@@ -1585,11 +1585,128 @@ func (s *PluginSuite) TestCreateRegistrationEntry() {
 		registrationEntry, err := s.ds.CreateRegistrationEntry(ctx, validRegistrationEntry)
 		s.Require().NoError(err)
 		s.Require().NotNil(registrationEntry)
-		s.NotEmpty(registrationEntry.EntryId)
-		registrationEntry.EntryId = ""
-		s.assertCreatedAtField(registrationEntry, now)
-		registrationEntry.CreatedAt = validRegistrationEntry.CreatedAt
-		s.RequireProtoEqual(registrationEntry, validRegistrationEntry)
+		s.assertEntryEqual(s.T(), validRegistrationEntry, registrationEntry, now)
+	}
+}
+
+func (s *PluginSuite) TestCreateOrReturnRegistrationEntry() {
+	now := time.Now().Unix()
+
+	for _, tt := range []struct {
+		name          string
+		updateEntry   func(*common.RegistrationEntry) *common.RegistrationEntry
+		expectError   string
+		expectSimilar bool
+	}{
+		{
+			name: "no entry provided",
+			updateEntry: func(e *common.RegistrationEntry) *common.RegistrationEntry {
+				return nil
+			},
+			expectError: "datastore-sql: invalid request: missing registered entry",
+		},
+		{
+			name: "no selectors",
+			updateEntry: func(e *common.RegistrationEntry) *common.RegistrationEntry {
+				e.Selectors = nil
+				return e
+			},
+			expectError: "datastore-sql: invalid registration entry: missing selector list",
+		},
+		{
+			name: "no SPIFFE ID",
+			updateEntry: func(e *common.RegistrationEntry) *common.RegistrationEntry {
+				e.SpiffeId = ""
+				return e
+			},
+			expectError: "datastore-sql: invalid registration entry: missing SPIFFE ID",
+		},
+		{
+			name: "negative X509 ttl",
+			updateEntry: func(e *common.RegistrationEntry) *common.RegistrationEntry {
+				e.X509SvidTtl = -1
+				return e
+			},
+			expectError: "datastore-sql: invalid registration entry: X509SvidTtl is not set",
+		},
+		{
+			name: "negative JWT ttl",
+			updateEntry: func(e *common.RegistrationEntry) *common.RegistrationEntry {
+				e.JwtSvidTtl = -1
+				return e
+			},
+			expectError: "datastore-sql: invalid registration entry: JwtSvidTtl is not set",
+		},
+		{
+			name: "create entry successfully",
+			updateEntry: func(e *common.RegistrationEntry) *common.RegistrationEntry {
+				return e
+			},
+		},
+		{
+			name: "subset selectors",
+			updateEntry: func(e *common.RegistrationEntry) *common.RegistrationEntry {
+				e.Selectors = []*common.Selector{
+					{Type: "a", Value: "1"},
+				}
+				return e
+			},
+		},
+		{
+			name: "with partial selectors",
```

Comment:
```suggestion
			name: "with superset selectors",
```
nit on naming

also for an actual test of `partial`, while we're doing sanity testing of this check, could do an intersection, `a:1` and `d:4` and a mutually exclusive `d:4` and `e:5`

Reply:
> also for an actual test of partial, while we're doing sanity testing of this check, could do an intersection, a:1 and d:4 and a mutually exclusive d:4 and e:5

would be nice to do this one still; we'll be a bit sad in the future if we break this functionality in an interesting way that's not caught by unit testing the possible Set permutations